docs/content/posts/2023-04-12-howto-migrate-to-datastream.md

```diff
@@ -0,0 +1,190 @@
+Collection. So we store in the reduced row the partially aggregated sum. Due to incremental reduce,
+we also need to distinguish if we received an already reduced row (in that case, we read the
+partially aggregated sum) or a fresh row (in that case we just read the corresponding price field).
+
+### [Migrating the order by operation](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java#L172-L199)
+
+The sort of the datastream is done by applying
+a [KeyedProcessFunction](https://javadoc.io/static/org.apache.flink/flink-streaming-java/1.16.0/org/apache/flink/streaming/api/functions/KeyedProcessFunction.html)
+.
+
+But, as said above, the DataStream semantics are the ones of a streaming pipeline. The arriving data
+is thus considered infinite. As such we need to "divide" the data to have output times. For that we
+need to set a timer to output the resulting data. As we are in batch mode we work in processing
+time (and not in event time)
+and [set a timer to fire at Long.MAX_VALUE timestamp](https://github.com/echauchot/tpcds-benchmark-flink/blob/9c65e535bbd7f9c7f507e499c31c9280be2993ca/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java#L179)
+meaning that the timer will fire at the end of the batch.
+
+To sort the data, we store the incoming rows inside
+a [ListState](https://javadoc.io/static/org.apache.flink/flink-core/1.16.0/org/apache/flink/api/common/state/ListState.html)
+and sort them at output time, when the timer fires in
+the [onTimer()](https://javadoc.io/static/org.apache.flink/flink-streaming-java/1.16.0/org/apache/flink/streaming/api/functions/KeyedProcessFunction.html#onTimer-long-org.apache.flink.streaming.api.functions.KeyedProcessFunction.OnTimerContext-org.apache.flink.util.Collector-)
+callback.
+
+Another thing: to be able to use Flink state, we need to key the datastream beforeward even if there
+is no group by key because Flink state is designed per-key. Thus, we key by a fake static key so
+that there is a single state.
+
+### [Migrating the limit operation](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java#L201-L204)
+
+Like for the order by migration, here also we need to adapt to the streaming semantics in which we
+don't have the whole data at hand. Once again, we use the state API. We apply a mapper that stores
+the incoming rows into
+a [ValueState](https://javadoc.io/static/org.apache.flink/flink-core/1.16.0/org/apache/flink/api/common/state/ValueState.html)
+to count them and stop at the set limit. Here also we need to key by a static key to be able to use
+the state API.
+The code resides in
+the [LimitMapper](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java#L232-L253)
```

Comment:
This seems weird. It's super sketchy for a function to just close() itself.

Why not map records to "1" and apply a reduce function that maxes out at 100?

Reply:
agree. I've found a bug in the code of this article, will fix and get back here. 

Also please note that I could have done much simpler code by mixing DataStream code with SQL/Table API especially using `LIMIT 100` and `ORDER BY` but the point was to try to code the equivalent of DataSet features with pure DataStream API to prepare the migration path for users (discover limitations such as [this one](https://issues.apache.org/jira/browse/FLINK-22587)) and also to compare the performance of both APIs.